File: 5. Token ring/TokenRing.java
import java.util.Scanner;

public class TokenRing {
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        int token = 0;

        System.out.print("Enter the No. of processes: ");
        int n = sc.nextInt();

        // Display ring
        System.out.print("Ring:");
        for(int i = 0; i < n; i++){
            System.out.print(" P" + i + " ->");
        }
        System.out.println(" P0");

        int s, r, data;

        while(true){
            System.out.println("\n  Token is at P" + token);
            System.out.print("  Enter the sender process (0-" + (n-1) + "): ");
            s = sc.nextInt();
            System.out.print("  Enter the receiver process (0-" + (n-1) + "): ");
            r = sc.nextInt();
            System.out.print("  Enter the data to send: ");
            data = sc.nextInt();

            System.out.print("\n  PHASE 1: [TOKEN Traversal]\n  ");
            while(token != s){
                System.out.print(" P" + token + " ->");
                token = (token + 1) % n;
            }
            System.out.println(" P" + token + " (sender receives token)");

            System.out.println("\n  PHASE 2: [DATA FORWARDING]");

            if(s == r){
                System.out.println("  Sender and Receiver are the same node!");
                System.out.println("  P" + s + " receives its own data: " + data);
            } else {
                System.out.print("  Data travels: P" + s);
                token = (token + 1) % n;
                while(token != r){
                    System.out.print(" -> P" + token);
                    token = (token + 1) % n;
                }
                System.out.println(" -> P" + token);
                System.out.println("\n  P" + r + " (receiver gets data): " + data);
            }
        }
    }
}

File: 5. Token ring/TokenRing.cpp
#include <bits/stdc++.h>
using namespace std;
int main(){
    int token = 0;
    int n;
    cout << "Enter the No. of processs: ";
    cin >> n;
    cout << "Ring: ";
    for (int i = 0; i < n; i++){
        cout << " P" << i << " ->";
    }
    cout << " P0" << endl;

    int s, r, data;

    while (true){
        cout << "\n  Token is at P" << token << endl;
        cout << "  Enter the sender process (0-" << (n - 1) << ") ";
        cin >> s;
        cout << "  Enter the Receiver process (0-" << (n - 1) << ") ";
        cin >> r;
        cout << "  Enter the data to send: ";
        cin >> data;
        cout << "\n  PHASE 1:[TOKEN Traversal]\n ";
        // if()
        while (token != s){
            cout << " P" << token << " ->";
            token = (token + 1) % n;
        }
        cout << " P" << token << " (sender receive token)" << endl;

        cout << "\n  PHASE 2:[DATA FORWARDING]\n";
        cout << "  sender forward data to\n  Node:";
        while (token != r){
            cout << " P" << token << " ->";
            token = (token + 1) % n;
        }
        cout << " P" << token << endl;
        cout << "\n  P" << token << " (reciever receive token and data) data: " << data << endl;
        cout << "Token is empty again!!!" << endl;
    }

    return 0;
}